### PR TITLE
WIP: Address #12074 - Date array operations

### PR DESCRIPTION
**Pull request by GordStephen**
A first step at improving functionality consistency among array operations involving `TimeType` and `Period`/`CompoundPeriod` (as per https://github.com/JuliaLang/julia/issues/12074). For now I've just implemented subtraction between `TimeType` arrays and scalars.

~~Unfortunately, defining~~

``` julia
(-){T<:TimeType}(x::AbstractArray{T}, y::AbstractArray{T})
```

~~raises ambiguous definition errors with both https://github.com/JuliaLang/julia/blob/master/base/arraymath.jl#L71 and https://github.com/JuliaLang/julia/blob/master/base/arraymath.jl#L80. Any suggestions to avoid that?~~
Edit: Fixed below

### CODE REVIEW

**Comment by nalimilan** [comment]
I think you should define these to get rid of the ambiguities:
```julia
(-){T<:TimeType}(x::AbstractArray{T}, y::Range{T})
(-){T<:TimeType}(x::Range{T}, y::AbstractArray{T})
```

**Comment by GordStephen** [comment]
Hmm, so that actually results in more ambiguity warnings:

From the first new method:
```julia
-(Base.Range{S<:Any}, Base.Range{T<:Any})
```
From the second:
```julia
-(Base.Range{S<:Any}, Base.Range{T<:Any})
-(AbstractArray{T<:Base.Dates.TimeType, N<:Any}, Base.Range{T<:Base.Dates.TimeType})
-(AbstractArray{S<:Any, N<:Any}, Base.Range{T<:Any})
```

Explicitly defining
```julia
(-){T<:TimeType}(Base.Range{T}, Base.Range{T})
```

doesn't seem to make any difference one way or the other.

**Comment by nalimilan** [comment]
Indeed, you also need to allow two different element types.
```julia
(-){S<:TimeType, T<:TimeType}(x::AbstractArray{S}, y::Range{T})
(-){S<:TimeType, T<:TimeType}(x::Range{S}, y::AbstractArray{T})
```

I'm not sure this makes sense, though. You may also have a method just to raise an error if it doesn't.

If you still get warnings, carefully look at the error message: everything is explained, though it's a bit hard to read.

**Comment by GordStephen** [comment]
The underlying scalar operation requires that the types match, so the vector version should as well, no?

```julia
(-){T<:TimeType}(x::T,y::T) = x.instant - y.instant
```

File: base/dates/arithmetic.jl [hunk 1]
```diff
@@ -68,5 +68,10 @@ end
 
 (.+){T<:TimeType}(x::AbstractArray{T}, y::Period) = reshape(T[i + y for i in x], size(x))
 (.-){T<:TimeType}(x::AbstractArray{T}, y::Period) = reshape(T[i - y for i in x], size(x))
-(.+){T<:TimeType}(y::Period, x::AbstractArray{T}) = x .+ y
-(.-){T<:TimeType}(y::Period, x::AbstractArray{T}) = x .- y
+(.+)(y::Period, x::AbstractArray{TimeType}) = x .+ y
```
Comment (tkelman) [inline base/dates/arithmetic.jl]: this changes the meaning since `TimeType` is abstract, you probably won't have arrays of them very often

Comment (GordStephen) [inline base/dates/arithmetic.jl]: Ah, good point, thanks.

**Comment by GordStephen** [comment]
#12115 addresses everything here.

**Pull request closed without merging**